### ISSUE TITLE
Fail doing integrated test

### ISSUE THREAD

**Issue by akapawat** [boutproject/BOUT-dev#1288]
I've just install BOUT++ for my incoming project Everything went fine until I run make check command. It showed 
>makefile:45: recipe for target 'check-integrated-tests' failed
>make: *** [check-integrated-tests] Error 1

I've struck on this issue for weeks now. I've tried reinstall BOUT++ and other dependencies but it doesn't work. If anyone know what I should do please help. Thanks

I'm running on Ubantu 18.04 on VirtualBox. and I've also attached my <make.config> file in text format if it will help
[make .txt](https://github.com/boutproject/BOUT-dev/files/2425028/make.txt)

.

**Comment by bendudson** [comment]
Hello, could you please attach the full error message? Which tests are failing and what is the full error message you see?

**Comment by akapawat** [comment]
Sorry for the late reply. I just had my hand back on the project. This is what I got when I ran <make check>. I dont know what should I put in so I copied all output into this text file.
[error.txt](https://github.com/boutproject/BOUT-dev/files/2428581/error.txt)

**Comment by dschwoerer** [comment]
They all seem to be of the sort:
Number of processors (1) not divisible by NPs in x direction (3)

I guess mpirun is not working.

What is the output of `mpirun -np 3 echo 1` ?

**Comment by dschwoerer** [comment]
That is actually strange, it seems mpirun is starting the appropriate number of jobs, but they are not communicating with each other ...

I assume Ubanyu means Ubuntu?

could you run this and post the output:
```
mpirun --version
which mpirun
which mpicc
uname -a
```

**Comment by akapawat** [comment]
It's Ubantu, sorry for the typo.
This is the output

```
mpirun --version
mpirun (Open MPI) 2.1.1

Report bugs to http://www.open-mpi.org/community/help/

which mpirun
/usr/bin/mpirun

which mpicc
/usr/bin/mpicc

uname -a
Linux pawataka-VirtualBox 4.15.0-34-generic #37-Ubuntu SMP Mon Aug 27 15:21:48 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by dschwoerer** [comment]
I have downloaded ubuntu 18.04 and installed in VirtualBox.
Host is an x86_64 and Guest is also 64 bit (this can be checked by running `arch` on host and guest)
After that, I run: (removed some unrelated commands, or stuff I canceled)
```
    5  sudo apt install git gcc mpi
    6  sudo apt-get install mpich libmpich-dev libfftw3-dev libnetcdf-dev libnetcdf-cxx-legacy-dev git g++ make git
   13  sudo apt-get remove mpich libmpich-dev 
   16  sudo apt install libblacs-mpi-dev
   17  mpirun --version
   20  git clone https://github.com/boutproject/BOUT-dev.git
   21  cd BOUT-dev/
   22  git checkout a8157290fedb1b589d0a0008e05b75611e4a8842
   24  ./configure 
   26  make 
   27  make check
   29  sudo apt-get install python3 python3-distutils python3-pip python3-numpy python3-netcdf4 python3-scipy
   30  make check
```
All succeeded.

Could you mention how you install openmpi?
I am not using ubuntu, so I don't know how to verify for package integrity ...

Further, could you compile this:
```
/* program hello */
/* Adapted from mpihello.f by drs */

#include <mpi.h>
#include <stdio.h>
#include <unistd.h>

int main(int argc, char **argv)
{
  int rank;
  char hostname[256];

  MPI_Init(&argc,&argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  gethostname(hostname,255);

  printf("Hello world!  I am process number: %d on host %s\n", rank, hostname);

  MPI_Finalize();

  return 0;
}
```
and run afterwards:
`mpirun -np 3 ./a.out`?

I guess that your openmpi installation is corrupted.

An easy fix would be to start with a new VirtualBox installation of ubuntu, and follow the instructions for ubuntu 18.04 [here](https://bout-dev.readthedocs.io/en/latest/user_docs/installing.html#ubuntu-debian)

**Comment by akapawat** [comment]
Guess you're right. I installed openmpi via the link in BOUT++ tutorial 
I can't compile your code; it said that there is no mpi.h directory

```
gcc test.c 
test.c:4:10: fatal error: mpi.h: No such file or directory
 #include <mpi.h>
          ^~~~~~~

compilation terminated.
```

neither can I run <mpirun -np 3 ./a.out>

```
mpirun -np 3 .a/.out
--------------------------------------------------------------------------
mpirun was unable to launch the specified application as it could not access
or execute an executable:

Executable: .a/.out
Node: pawataka-VirtualBox

while attempting to start process rank 0.
--------------------------------------------------------------------------
3 total processes failed to start`
```

Is there other way to fix my mpi without reinstalling my ubantu?

**Comment by d7919** [comment]
I think you need to use mpicc rather than GCC to try to compile the example.

You could try removing the openmpi package (maybe by sudo apt-get remove
openmpi) and then reinstalling (sudo apt-get install openmpi or similar)
bit this might have further side effects.

**Comment by akapawat** [comment]
Ok. I typed <mpicc test.c> but nothing happened. Did I do it correctly?

**Comment by dschwoerer** [comment]
Yes, if there is no error message, most likely no error occurred.
As a side note, you can check for the error code by running `echo $?` just after the program finished - it should print a `0` if everything went fine.

After you compiled successfully, could you try again running the program with:
```
mpirun -np 3 .a/.out
```

**Comment by dschwoerer** [comment]
As I thought, this is a bug in your OpenMPI installation, rather than in BOUT++.

How did you install OpenMPI? Could you provide a link to the instructions?

One option would be to reinstall OpenMPI, think this might work:
```
sudo apt remove libopenmpi2 openmpi-common
sudo apt install libopenmpi-dev
```

Another option is switching to mpich:
```
sudo apt-get install mpich libmpich-dev
# Verify both openmpi and mpich are available - should print both lines
update-alternatives --list mpi
# Select mpich:
sudo update-alternatives --set mpi /usr/include/mpich
# Verify mpicc points to mpich:
readlink /usr/bin/mpicc -f
```

After that, make sure that the above example compiles, and prints the numbers 0 to 2, when run. After that, try to compile BOUT++ again.

Another option would be to get in contact with the OpenMPI developer, they might be able to give more insight what is wrong with your setup ...

**Comment by akapawat** [comment]
I'm not sure how I got my openmpi installed. All I remembered was I got it from some website listed on BOUT++ tutorial site.
Anyway, so this is the output I received after I reinstalled openmpi
```
mpicc test.c 
echo $?
mpirun -np 3 ./a.out
Hello world!  I am process number: 1 on host pawataka-VirtualBox
Hello world!  I am process number: 2 on host pawataka-VirtualBox
Hello world!  I am process number: 0 on host pawataka-VirtualBox
```
Also, I compiled BOUT++ again  and every tests went fine. Now I'm going to run some calculation examples and see if it works.
I think this is it for now. Thank you for your time and your help. I really appreciate that.
